### PR TITLE
feat(agentkit): Add Uniswap V3 Action Provider for Base Network

### PR DESCRIPTION
**Pull request by Thedongraphix**
## Summary

This PR introduces a new **Uniswap V3 Action Provider** for the Base network, enabling AI agents to perform token swaps on one of the most popular DEXs. This addresses a fundamental need for onchain agents and directly responds to functionality mentioned in the project's WISHLIST.md.

## Motivation

Token swaps are a core primitive for DeFi operations. Every onchain agent needs the ability to exchange tokens, whether for:
- Portfolio rebalancing
- Yield farming strategies  
- Arbitrage opportunities
- Basic token acquisition

This provider makes it simple for agents to get quotes and execute swaps with a single tool call.

### 🔄 **Two Core Actions**
- **`get_quote`**: Get swap quotes without executing (for planning & price discovery)
- **`swap`**: Execute token swaps with comprehensive safety features

### 🛡️ **Built-in Safety & UX**
- **Slippage Protection**: Configurable slippage tolerance (default 0.5%)
- **Automatic ERC20 Approvals**: Handles token allowances automatically
- **Native ETH Support**: Seamlessly works with WETH for ETH swaps
- **Clear Error Messages**: Helpful feedback for troubleshooting

### 🌐 **Network Support**
- Base Mainnet
- Base Sepolia (for testing)

### Files Added
typescript/agentkit/src/action-providers/uniswap/
├── constants.ts # Contract addresses & ABIs
├── schemas.ts # Input validation schemas
├── uniswapActionProvider.ts # Main implementation
├── index.ts # Exports
└── tests/
└── uniswapActionProvider.test.ts # Unit tests

### Architecture
- **Modular Design**: Self-contained provider following agentkit patterns
- **Type Safety**: Full TypeScript support with Zod validation
- **Consistent API**: Uses same patterns as existing providers (ERC20, Compound, etc.)
- **Extensible**: Easy to add more Uniswap features in the future

## Usage Example

```typescript
// Get a quote
const quote = await agent.getQuote({
  tokenIn: "0x...", // USDC address
  tokenOut: "0x...", // WETH address  
  amount: "100",    // 100 USDC
  fee: 3000        // 0.3% fee tier
});

// Execute swap
const result = await agent.swap({
  tokenIn: "0x...",
  tokenOut: "0x...",
  amount: "100",
  slippageTolerance: 0.5, // 0.5%
  fee: 3000
});
```

## Testing

- ✅ TypeScript compilation passes (`npm run build`)
- ✅ Unit tests verify network support
- ✅ Follows existing test patterns
- ✅ Proper error handling

## Value Proposition

This provider:
- **Solves a core need**: Every agent needs token swaps
- **Reduces complexity**: Handles approvals, slippage, and edge cases automatically  
- **Follows best practices**: Uses established agentkit patterns
- **Enables innovation**: Unblocks agents from building advanced DeFi strategies

## Breaking Changes

None - this is a purely additive change.

## Additional Context

This implementation focuses on the most commonly used Uniswap V3 functionality while maintaining simplicity and safety. Future enhancements could include:
- Multi-hop swaps
- Liquidity provision
- Fee tier optimization
- Advanced order types

---

**Ready to empower AI agents with seamless token swapping on Base!**

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td><code>2</code> if repo is sensitive</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>1</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>2</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

**Comment by phdargen** [comment]
Hi @Thedongraphix, thanks for your contribution! Looks pretty good, a few initial comments below.

- please move uniswapActionProvider.test.ts directly into uniswap provider directory to be consistent with existing project file structure

- add a readme in uniswap provider directory
- add entry in agentkit/README (https://github.com/coinbase/agentkit/blob/main/typescript/agentkit/README.md), should follow alphabetic order
 
- add a changeset, see https://github.com/coinbase/agentkit/blob/main/CONTRIBUTING-TYPESCRIPT.md

- please run pnpm run format and pnpm run lint before you commit

- please also verify your commit, see https://docs.github.com/en/authentication/managing-commit-signature-verification

File: typescript/agentkit/src/action-providers/uniswap/schemas.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import { z } from "zod";
+
+/**
+ * Input schema for Uniswap token swap action.
+ */
+export const SwapSchema = z
+  .object({
+    tokenIn: z.string().describe("The contract address of the token to swap from"),
```
Comment (phdargen) [inline typescript/agentkit/src/action-providers/uniswap/schemas.ts:8]: suggest to use `.regex(/^0x[a-fA-F0-9]{40}$/, "Invalid Ethereum address format"),` for all contract addresses

File: typescript/agentkit/src/action-providers/uniswap/__tests__/uniswapActionProvider.test.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { uniswapActionProvider } from "../uniswapActionProvider";
+
+describe("UniswapActionProvider", () => {
+  const actionProvider = uniswapActionProvider();
```
Comment (phdargen) [inline typescript/agentkit/src/action-providers/uniswap/__tests__/uniswapActionProvider.test.ts:4]: Please add tests for the logic of each action. As a guide you can have a look at eg https://github.com/coinbase/agentkit/blob/main/typescript/agentkit/src/action-providers/flaunch/flaunchActionProvider.test.ts that also tests some swap actions

**Comment by phdargen** [comment]
It would be good to include the result of some manual end-to-end testing in the PR description. Add the new uniswap action provider to one of the chatbot examples, run it and copy your prompts + output for quote and swap queries here.

Please keep changes to the chatbot examples local and do not commit them.

File: typescript/agentkit/src/action-providers/uniswap/constants.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
+import type { Abi } from "abitype";
+
+/**
+ * Uniswap V3 contract addresses for Base networks
+ */
+export const UNISWAP_ADDRESSES: Record<string, Record<string, string>> = {
```
Comment (phdargen) [inline typescript/agentkit/src/action-providers/uniswap/constants.ts:6]: Optional: could be extended to a few more chains like ethereum, optimism, arbitrum, polygon (https://github.com/coinbase/agentkit/blob/main/typescript/agentkit/src/network/network.ts)

You can find the contract addresses here:
https://docs.uniswap.org/contracts/v3/reference/deployments/
